// react-data-grid/lib/Cell.js
import React, { forwardRef, memo } from 'react';
import clsx from 'clsx';
import { preventDefault, wrapEvent } from './utils';
function Cell({ className, column, isRowSelected, lastFrozenColumnIndex, row, rowIdx, eventBus, onRowClick, onClick, onDoubleClick, onContextMenu, onDragOver, ...props }, ref) {
    function selectCell(openEditor) {
        eventBus.dispatch('SELECT_CELL', { idx: column.idx, rowIdx }, openEditor);
    }
    function handleCellClick() {
        selectCell();
        onRowClick?.(rowIdx, row, column);
    }
    function handleCellContextMenu() {
        selectCell();
    }
    function handleCellDoubleClick() {
        selectCell(true);
    }
    function onRowSelectionChange(checked, isShiftClick) {
        eventBus.dispatch('SELECT_ROW', { rowIdx, checked, isShiftClick });
    }
    const { cellClass } = column;
    className = clsx('rdg-cell', {
        'rdg-cell-frozen': column.frozen,
        'rdg-cell-frozen-last': column.idx === lastFrozenColumnIndex
    }, typeof cellClass === 'function' ? cellClass(row) : cellClass, className);
    return (React.createElement("div", Object.assign({ ref: ref, className: className, style: {
            width: column.width,
            left: column.left
        }, onClick: wrapEvent(handleCellClick, onClick), onDoubleClick: wrapEvent(handleCellDoubleClick, onDoubleClick), onContextMenu: wrapEvent(handleCellContextMenu, onContextMenu), onDragOver: wrapEvent(preventDefault, onDragOver) }, props),
        React.createElement(column.formatter, { column: column, rowIdx: rowIdx, row: row, isRowSelected: isRowSelected, onRowSelectionChange: onRowSelectionChange })));
}
export default memo(forwardRef(Cell));
