// react-data-grid/lib/DataGrid.js
import React, { forwardRef, useState, useRef, useMemo, useLayoutEffect, useEffect, useImperativeHandle, useCallback, createElement } from 'react';
import EventBus from './EventBus';
import InteractionMasks from './masks/InteractionMasks';
import HeaderRow from './HeaderRow';
import FilterRow from './FilterRow';
import Row from './Row';
import SummaryRow from './SummaryRow';
import { ValueFormatter } from './formatters';
import { assertIsValidKey, getColumnMetrics, getColumnScrollPosition, getHorizontalRangeToRender, getScrollbarSize, getVerticalRangeToRender, getViewportColumns } from './utils';
import { CellNavigationMode } from './common/enums';
/**
 * Main API Component to render a data grid of rows and columns
 *
 * @example
 *
 * <DataGrid columns={columns} rows={rows} />
*/
function DataGrid({ 
// Grid and data Props
columns: rawColumns, rows, summaryRows, rowKey, onRowsUpdate, 
// Dimensions props
width, height = 350, minColumnWidth = 80, rowHeight = 35, headerRowHeight = rowHeight, headerFiltersHeight = 45, 
// Feature props
selectedRows, onSelectedRowsChange, sortColumn, sortDirection, onSort, filters, onFiltersChange, 
// Custom renderers
defaultFormatter = ValueFormatter, rowRenderer: RowRenderer = Row, emptyRowsRenderer, 
// Event props
onRowClick, onScroll, onColumnResize, onSelectedCellChange, onCheckCellIsEditable, 
// Toggles and modes
enableFilters = false, enableCellAutoFocus = true, enableCellCopyPaste = false, enableCellDragAndDrop = false, cellNavigationMode = CellNavigationMode.NONE, 
// Miscellaneous
editorPortalTarget = document.body, rowClass }, ref) {
    /**
     * refs
     * */
    const gridRef = useRef(null);
    const lastSelectedRowIdx = useRef(-1);
    /**
     * states
     */
    const [eventBus] = useState(() => new EventBus());
    const [gridWidth, setGridWidth] = useState(0);
    const [scrollTop, setScrollTop] = useState(0);
    const [scrollLeft, setScrollLeft] = useState(0);
    const [columnWidths, setColumnWidths] = useState(() => new Map());
    /**
     * computed values
     */
    const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
    const { columns, lastFrozenColumnIndex, totalColumnWidth } = useMemo(() => {
        return getColumnMetrics({
            columns: rawColumns,
            minColumnWidth,
            viewportWidth,
            columnWidths,
            defaultFormatter
        });
    }, [columnWidths, rawColumns, defaultFormatter, minColumnWidth, viewportWidth]);
    const [colOverscanStartIdx, colOverscanEndIdx] = useMemo(() => {
        return getHorizontalRangeToRender(columns, lastFrozenColumnIndex, viewportWidth, scrollLeft);
    }, [scrollLeft, columns, lastFrozenColumnIndex, viewportWidth]);
    const viewportColumns = useMemo(() => {
        return getViewportColumns(columns, colOverscanStartIdx, colOverscanEndIdx);
    }, [colOverscanEndIdx, colOverscanStartIdx, columns]);
    const totalHeaderHeight = headerRowHeight + (enableFilters ? headerFiltersHeight : 0);
    const clientHeight = height
        - 2 // border width
        - totalHeaderHeight
        - (summaryRows?.length ?? 0) * rowHeight
        - (totalColumnWidth > viewportWidth ? getScrollbarSize() : 0);
    const [rowOverscanStartIdx, rowOverscanEndIdx] = getVerticalRangeToRender(clientHeight, rowHeight, scrollTop, rows.length);
    /**
     * effects
     */
    useLayoutEffect(() => {
        // Do not calculate the width if width is provided
        if (typeof width === 'number')
            return;
        function onResize() {
            // Immediately re-render when the component is mounted to get valid columnMetrics.
            setGridWidth(gridRef.current.getBoundingClientRect().width);
        }
        onResize();
        window.addEventListener('resize', onResize);
        return () => {
            window.removeEventListener('resize', onResize);
        };
    }, [width]);
    useEffect(() => {
        if (!onSelectedRowsChange)
            return;
        const handleRowSelectionChange = ({ rowIdx, checked, isShiftClick }) => {
            assertIsValidKey(rowKey);
            const newSelectedRows = new Set(selectedRows);
            const rowId = rows[rowIdx][rowKey];
            if (checked) {
                newSelectedRows.add(rowId);
                const previousRowIdx = lastSelectedRowIdx.current;
                lastSelectedRowIdx.current = rowIdx;
                if (isShiftClick && previousRowIdx !== -1 && previousRowIdx !== rowIdx) {
                    const step = Math.sign(rowIdx - previousRowIdx);
                    for (let i = previousRowIdx + step; i !== rowIdx; i += step) {
                        newSelectedRows.add(rows[i][rowKey]);
                    }
                }
            }
            else {
                newSelectedRows.delete(rowId);
                lastSelectedRowIdx.current = -1;
            }
            onSelectedRowsChange(newSelectedRows);
        };
        return eventBus.subscribe('SELECT_ROW', handleRowSelectionChange);
    }, [eventBus, onSelectedRowsChange, rows, rowKey, selectedRows]);
    useImperativeHandle(ref, () => ({
        scrollToColumn(idx) {
            scrollToCell({ idx });
        },
        scrollToRow(rowIdx) {
            const { current } = gridRef;
            if (!current)
                return;
            current.scrollTop = rowIdx * rowHeight;
        },
        selectCell(position, openEditor) {
            eventBus.dispatch('SELECT_CELL', position, openEditor);
        }
    }));
    /**
     * event handlers
     */
    function onGridScroll(event) {
        const { scrollTop, scrollLeft } = event.currentTarget;
        setScrollTop(scrollTop);
        setScrollLeft(scrollLeft);
        onScroll?.(event);
    }
    const handleColumnResize = useCallback((column, width) => {
        const newColumnWidths = new Map(columnWidths);
        newColumnWidths.set(column.key, width);
        setColumnWidths(newColumnWidths);
        onColumnResize?.(column.idx, width);
    }, [columnWidths, onColumnResize]);
    function handleRowsUpdate(event) {
        onRowsUpdate?.(event);
    }
    /**
     * utils
     */
    function getFrozenColumnsWidth() {
        if (lastFrozenColumnIndex === -1)
            return 0;
        const lastFrozenCol = columns[lastFrozenColumnIndex];
        return lastFrozenCol.left + lastFrozenCol.width;
    }
    function scrollToCell({ idx, rowIdx }) {
        const { current } = gridRef;
        if (!current)
            return;
        if (typeof idx === 'number' && idx > lastFrozenColumnIndex) {
            const { clientWidth } = current;
            const { left, width } = columns[idx];
            const isCellAtLeftBoundary = left < scrollLeft + width + getFrozenColumnsWidth();
            const isCellAtRightBoundary = left + width > clientWidth + scrollLeft;
            if (isCellAtLeftBoundary || isCellAtRightBoundary) {
                const newScrollLeft = getColumnScrollPosition(columns, idx, scrollLeft, clientWidth);
                current.scrollLeft = scrollLeft + newScrollLeft;
            }
        }
        if (typeof rowIdx === 'number') {
            if (rowIdx * rowHeight < scrollTop) {
                // at top boundary, scroll to the row's top
                current.scrollTop = rowIdx * rowHeight;
            }
            else if ((rowIdx + 1) * rowHeight > scrollTop + clientHeight) {
                // at bottom boundary, scroll the next row's top to the bottom of the viewport
                current.scrollTop = (rowIdx + 1) * rowHeight - clientHeight;
            }
        }
    }
    function getViewportRows() {
        const rowElements = [];
        for (let rowIdx = rowOverscanStartIdx; rowIdx <= rowOverscanEndIdx; rowIdx++) {
            const row = rows[rowIdx];
            let key = rowIdx;
            let isRowSelected = false;
            if (rowKey !== undefined) {
                const rowId = row[rowKey];
                isRowSelected = selectedRows?.has(rowId) ?? false;
                if (typeof rowId === 'string' || typeof rowId === 'number') {
                    key = rowId;
                }
            }
            rowElements.push(React.createElement(RowRenderer, { key: key, rowIdx: rowIdx, row: row, viewportColumns: viewportColumns, lastFrozenColumnIndex: lastFrozenColumnIndex, eventBus: eventBus, isRowSelected: isRowSelected, onRowClick: onRowClick, rowClass: rowClass, top: rowIdx * rowHeight + totalHeaderHeight }));
        }
        return rowElements;
    }
    return (React.createElement("div", { className: "rdg", style: {
            width,
            height,
            '--header-row-height': `${headerRowHeight}px`,
            '--filter-row-height': `${headerFiltersHeight}px`,
            '--row-width': `${totalColumnWidth}px`,
            '--row-height': `${rowHeight}px`
        }, ref: gridRef, onScroll: onGridScroll },
        React.createElement(HeaderRow, { rowKey: rowKey, rows: rows, columns: viewportColumns, onColumnResize: handleColumnResize, lastFrozenColumnIndex: lastFrozenColumnIndex, allRowsSelected: selectedRows?.size === rows.length, onSelectedRowsChange: onSelectedRowsChange, sortColumn: sortColumn, sortDirection: sortDirection, onSort: onSort }),
        enableFilters && (React.createElement(FilterRow, { lastFrozenColumnIndex: lastFrozenColumnIndex, columns: viewportColumns, filters: filters, onFiltersChange: onFiltersChange })),
        rows.length === 0 && emptyRowsRenderer ? createElement(emptyRowsRenderer) : (React.createElement(React.Fragment, null,
            viewportWidth > 0 && (React.createElement(InteractionMasks, { rows: rows, rowHeight: rowHeight, columns: columns, enableCellAutoFocus: enableCellAutoFocus, enableCellCopyPaste: enableCellCopyPaste, enableCellDragAndDrop: enableCellDragAndDrop, cellNavigationMode: cellNavigationMode, eventBus: eventBus, gridRef: gridRef, totalHeaderHeight: totalHeaderHeight, scrollLeft: scrollLeft, scrollTop: scrollTop, scrollToCell: scrollToCell, editorPortalTarget: editorPortalTarget, onCheckCellIsEditable: onCheckCellIsEditable, onRowsUpdate: handleRowsUpdate, onSelectedCellChange: onSelectedCellChange })),
            React.createElement("div", { style: { height: Math.max(rows.length * rowHeight, clientHeight) } }),
            getViewportRows(),
            summaryRows?.map((row, rowIdx) => (React.createElement(SummaryRow, { key: rowIdx, rowIdx: rowIdx, row: row, bottom: rowHeight * (summaryRows.length - 1 - rowIdx), viewportColumns: viewportColumns, lastFrozenColumnIndex: lastFrozenColumnIndex })))))));
}
export default forwardRef(DataGrid);
