// react-data-grid/lib/headerCells/ResizableHeaderCell.js
import React, { cloneElement } from 'react';
export default function ResizableHeaderCell({ children, column, onResize }) {
    function onMouseDown(event) {
        if (event.button !== 0) {
            return;
        }
        const { currentTarget } = event;
        const { right } = currentTarget.getBoundingClientRect();
        const offset = right - event.clientX;
        if (offset > 11) { // +1px to account for the border size
            return;
        }
        const onMouseMove = (event) => {
            handleResize(event.clientX + offset, currentTarget);
        };
        const onMouseUp = () => {
            window.removeEventListener('mousemove', onMouseMove);
            window.removeEventListener('mouseup', onMouseUp);
        };
        event.preventDefault();
        window.addEventListener('mousemove', onMouseMove);
        window.addEventListener('mouseup', onMouseUp);
    }
    function onTouchStart(event) {
        const touch = event.changedTouches[0];
        const { identifier } = touch;
        const { currentTarget } = event;
        const { right } = currentTarget.getBoundingClientRect();
        const offset = right - touch.clientX;
        if (offset > 11) { // +1px to account for the border size
            return;
        }
        function getTouch(event) {
            for (const touch of event.changedTouches) {
                if (touch.identifier === identifier)
                    return touch;
            }
            return null;
        }
        const onTouchMove = (event) => {
            const touch = getTouch(event);
            if (touch) {
                handleResize(touch.clientX + offset, currentTarget);
            }
        };
        const onTouchEnd = (event) => {
            const touch = getTouch(event);
            if (!touch)
                return;
            window.removeEventListener('touchmove', onTouchMove);
            window.removeEventListener('touchend', onTouchEnd);
        };
        window.addEventListener('touchmove', onTouchMove);
        window.addEventListener('touchend', onTouchEnd);
    }
    function handleResize(x, target) {
        const width = x - target.getBoundingClientRect().left;
        if (width > 0) {
            onResize(column, width);
        }
    }
    return cloneElement(children, {
        onMouseDown,
        onTouchStart,
        children: (React.createElement(React.Fragment, null,
            children.props.children,
            React.createElement("div", { className: "rdg-header-cell-resizer" })))
    });
}
